kentikapi/client_integration_test.go

```diff
@@ -426,83 +427,179 @@ func durationPtr(v time.Duration) *time.Duration {
-			codes.OK,
-			"",
+			name:                "request with invalid id, status InvalidArgument received",
+			request:             &syntheticspb.GetAgentRequest{Id: "1000"},
+			expectedRequestBody: "id:\"1000\"",
+			responses: []*gRPCResponse{
+				{status.Errorf(codes.NotFound, "Could not find synthetic agent id=1000"), nil},
+			},
+			expectedResult:    &syntheticspb.GetAgentResponse{},
+			expectedErrorCode: codes.NotFound,
+			expectedErrorMsg:  "Could not find synthetic agent id=1000",
+			expectedError:     true,
+		},
+		{
+			name:                "valid request, agent returned",
+			request:             &syntheticspb.GetAgentRequest{Id: "968"},
+			expectedRequestBody: "id:\"968\"",
+			responses: []*gRPCResponse{
+				{
+					status.Errorf(
+						codes.OK,
+						"OK",
+					),
+					&syntheticspb.GetAgentResponse{Agent: newDummyAgentForGRPC()},
+				},
+			},
+			expectedResult:    &syntheticspb.GetAgentResponse{Agent: newDummyAgentForGRPC()},
+			expectedErrorCode: codes.OK,
+			expectedErrorMsg:  "",
+			expectedError:     false,
 		},
 	}
 
-	ctx := context.Background()
-
-	lis, err := net.Listen("tcp", fmt.Sprintf("localhost:%d", port))
-	if err != nil {
-		log.Fatalf("failed to listen: %v", err)
-	}
-	s := grpc.NewServer()
-	pb.RegisterSyntheticsAdminServiceServer(s, &mockSyntheticsServer{})
-	go func() {
-		if err := s.Serve(lis); err != nil {
-			log.Fatalf("failed to serve: %v", err)
-		}
-	}()
-
-	client := kentikapi.NewClient(kentikapi.Config{
-		SyntheticsGRPCAPIURL: lis.Addr().String(),
-	})
-
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			request := &pb.GetAgentRequest{Id: tt.id}
+			s := newSpySyntheticsServer(t, tt.responses)
+			s.Start()
+			defer s.Stop()
+
+			client, err := kentikapi.NewClient(kentikapi.Config{
+				SyntheticsGRPCHostPort: s.url,
+				AuthToken:              dummyAuthToken,
+				AuthEmail:              dummyAuthEmail,
+				DisableTLS:             true,
+			})
+			require.NoError(t, err)
+
+			var header metadata.MD
+			response, err := client.SyntheticsAdmin.GetAgent(
+				context.Background(),
+				tt.request,
+				grpc.Header(&header),
+			)
 
-			response, err := client.GRPCSyntheticsAdminServiceAPI.GetAgent(ctx, request)
+			if tt.expectedError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
 
-			if response != nil {
-				assert.Equal(t, tt.res.GetAgent(), response.GetAgent())
+			if e, ok := status.FromError(err); ok {
+				assert.Equal(t, tt.expectedErrorCode, e.Code())
+				assert.Equal(t, tt.expectedErrorMsg, e.Message())
 			}
 
-			if err != nil {
-				assert.Errorf(t, err, tt.errMsg)
+			assert.Equal(t, len(s.responses), len(s.requests), "invalid number of requests")
+			for _, r := range s.requests {
+				assert.Equal(t, dummyAuthEmail, header.Get(authEmailKey)[0])
+				assert.Equal(t, dummyAuthToken, header.Get(authAPITokenKey)[0])
+				assert.Equal(t, tt.expectedRequestBody, r.String())
```

Comment:
Did you consider comparing r.GetID() of requests instead of string representation? This would match the way we compare responses (we call .GetAgent() there instead of String())
It would make writing test cases easier and increase readability

Reply:
I wasn't aware of the GetID() method. I will adjust the code.

Reply:
Done